crates/primitives/src/receipt.rs

```diff
@@ -44,6 +48,95 @@ impl Receipt {
     }
 }
 
+/// A collection of receipts organized as a two-dimensional vector.
+#[derive(Clone, Debug, PartialEq, Eq, Default)]
+pub struct Receipts {
+    /// A two-dimensional vector of optional `Receipt` instances.
+    pub receipt_vec: Vec<Vec<Option<Receipt>>>,
+}
+
+impl Receipts {
+    /// Create a new `Receipts` instance with an empty vector.
+    pub fn new() -> Self {
+        Self { receipt_vec: vec![] }
+    }
+
+    /// Create a new `Receipts` instance from an existing vector.
+    pub fn from_vec(vec: Vec<Vec<Option<Receipt>>>) -> Self {
+        Self { receipt_vec: vec }
+    }
+
+    /// Returns the length of the `Receipts` vector.
+    pub fn len(&self) -> usize {
+        self.receipt_vec.len()
+    }
+
+    /// Returns `true` if the `Receipts` vector is empty.
+    pub fn is_empty(&self) -> bool {
+        self.receipt_vec.is_empty()
+    }
+
+    /// Push a new vector of receipts into the `Receipts` collection.
+    pub fn push(&mut self, receipts: Vec<Option<Receipt>>) {
+        self.receipt_vec.push(receipts);
+    }
+
+    /// Retrieves the receipt root for all recorded receipts from index.
+    pub fn root_slow(&self, index: usize) -> Option<H256> {
+        Some(calculate_receipt_root_ref(
+            &self.receipt_vec[index].iter().map(Option::as_ref).collect::<Option<Vec<_>>>()?,
```

Comment:
you might be able to avoid the `collect` here? calculate_receipt_root shoudl be able to operate over iterators

Reply:
This will just transform into references on receipts vec and check if all Receipt are present/Some. Inside would still be casted with ReceiptWithBloom, so this seems okay.

Reply:
@rakita So this seems ok to keep it like that using the `collect` no? To keep the same formalism as what we had before adding the `Receipts` type no?